binders/kafka-binder/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderMetrics.java

```diff
@@ -141,45 +142,35 @@ public void bindTo(MeterRegistry registry) {
 			String topic = topicInfo.getKey();
 			String group = topicInfo.getValue().getConsumerGroup();
 
-			final Gauge register = Gauge.builder(OFFSET_LAG_METRIC_NAME, this,
-					(o) -> computeAndGetUnconsumedMessages(topic, group)).tag("group", group)
-					.tag("topic", topic)
-					.description("Unconsumed messages for a particular group and topic")
-					.register(registry);
+			ToDoubleFunction<KafkaBinderMetrics> offsetComputation = computeOffsetComputationFunction(topic, group);
+			final Gauge register = Gauge.builder(OFFSET_LAG_METRIC_NAME, this, offsetComputation)
+				.tag("group", group)
+				.tag("topic", topic)
+				.description("Unconsumed messages for a particular group and topic")
+				.register(registry);
 
 			if (!(register instanceof NoopGauge)) {
 				//Schedule a task to compute the unconsumed messages for this group/topic every minute.
-				this.scheduler.scheduleWithFixedDelay(computeUnconsumedMessagesRunnable(topic, group, this.metadataConsumers),
-						10, DELAY_BETWEEN_TASK_EXECUTION, TimeUnit.SECONDS);
+				this.scheduler.scheduleWithFixedDelay(
+					() -> computeAndGetUnconsumedMessages(topic, group),
+			10,
+					DELAY_BETWEEN_TASK_EXECUTION,
+					TimeUnit.SECONDS
+				);
 			}
 		}
 	}
 
-	private Runnable computeUnconsumedMessagesRunnable(String topic, String group, Map<String, Consumer<?, ?>> metadataConsumers) {
-		return () -> {
-			try {
-				long lag = findTotalTopicGroupLag(topic, group, this.metadataConsumers);
-				this.unconsumedMessages.put(topic + "-" + group, lag);
-			}
-			catch (Exception ex) {
-				LOG.debug("Cannot generate metric for topic: " + topic, ex);
-			}
-		};
+	private ToDoubleFunction<KafkaBinderMetrics> computeOffsetComputationFunction(String topic, String group) {
+		if (this.binderConfigurationProperties.getMetrics().isRealtimeOffsetLagEnabled()) {
+			return (o) -> computeAndGetUnconsumedMessagesWithTimeout(topic, group);
+		} else {
+			return (o) -> this.unconsumedMessages.getOrDefault(topic + "-" + group, 0L);
+		}
 	}
 
-	private long computeAndGetUnconsumedMessages(String topic, String group) {
-		ExecutorService exec = Executors.newCachedThreadPool();
-		Future<Long> future = exec.submit(() -> {
-
-			long lag = 0;
-			try {
-				lag = findTotalTopicGroupLag(topic, group, this.metadataConsumers);
-			}
-			catch (Exception ex) {
-				LOG.debug("Cannot generate metric for topic: " + topic, ex);
-			}
-			return lag;
-		});
+	private long computeAndGetUnconsumedMessagesWithTimeout(String topic, String group) {
+		Future<Long> future = scheduler.submit(() -> computeAndGetUnconsumedMessages(topic, group));
```

Comment:
I am re-using the `scheduler` here because I do not think we need another thread-pool here. Any thoughts?